pipelines/matrix/src/matrix/pipelines/matrix_generation/nodes.py

```diff
@@ -42,144 +40,147 @@ def enrich_embeddings(
-        clinical_trials: Pairs dataset representing outcomes of recent clinical trials.
-        off_label: Pairs dataset representing off label drug disease uses.
-        orchard: Pairs dataset representing orchard feedback data.
+        graph_nodes: pyspark dataframe containing node embeddings.
+        clinical_trials: pyspark dataframe representing outcomes of recent clinical trials.
+        off_label: pyspark dataframe representing off label drug disease uses.
+        orchard: pyspark dataframe representing orchard feedback data.
 
     Returns:
-        Pairs dataframe containing all combinations of drugs and diseases that do not lie in the training set.
+        pyspark dataframe containing all combinations of drugs and diseases that do not lie in the training set with flags
     """
-    # Collect list of drugs and diseases
-    diseases_lst = diseases["id"].tolist()
-    # This try/except is to make modelling_run pipeline compatible with old drug list (pre-migration)
-    try:
-        drugs_df = drugs[["id", "ec_id"]]
-        drugs_column_remapping = {"ec_id": "ec_drug_id", "id": "source"}
-    except KeyError:
-        logger.warning("ec_id column not found in drugs dataframe; using id column instead")
-        drugs_column_remapping = {"id": "source"}
-        drugs_df = drugs[["id"]]
-
-    # Remove duplicates
-    drugs_df = drugs_df.drop_duplicates()
-    diseases_lst = list(set(diseases_lst))
-
-    # Remove drugs and diseases without embeddings
-    nodes_with_embeddings = set(graph._nodes["id"])
-    drugs_df = drugs_df[drugs_df["id"].isin(nodes_with_embeddings)]
-    diseases_lst = [disease for disease in diseases_lst if disease in nodes_with_embeddings]
-
-    # Generate all drug disease combinations
-    matrix_slices = []
-    for disease in tqdm(diseases_lst):
-        matrix_slice = pd.DataFrame(drugs_df.rename(drugs_column_remapping, axis=1).assign(target=disease))
-        matrix_slices.append(matrix_slice)
-
-    # Concatenate all slices at once
-    matrix = pd.concat(matrix_slices, ignore_index=True)
-
-    # Remove training set
-    train_pairs = known_pairs[~known_pairs["split"].eq("TEST")]
-    train_pairs_set = set(zip(train_pairs["source"], train_pairs["target"]))
-    is_in_train = matrix.apply(lambda row: (row["source"], row["target"]) in train_pairs_set, axis=1)
-    matrix = matrix[~is_in_train]
-    # Add flag columns for known positives and negatives
-    matrix = _add_flag_columns(matrix, known_pairs, clinical_trials, off_label, orchard)
+
+    def get_join_columns(left_columns: list[str], right_columns: list[str]) -> list[str]:
+        if "source_ec_id" in left_columns and "source_ec_id" in right_columns:
+            return ["source_ec_id", "target"]
+        else:
+            return ["source", "target"]
+
+    # 1. Filter out drugs and diseases without embeddings
+    drugs_in_graph = drugs.alias("drugs").join(node_embeddings, on="id", how="inner")
```

Comment:
Nitpick: Avoid lugging around the embeddings
```suggestion
    drugs_in_graph = drugs.alias("drugs").join(node_embeddings.select("id"), on="id", how="inner")
```